Send bank statement descriptors to payment intents

cesarcosta99 opened the pull request:
<!--
Did I add a title? A descriptive, yet concise, title.
-->

<!--
Issue: Link to the GitHub issue this PR addresses (if appropriate).
-->

Fixes #2205 

## Changes proposed in this Pull Request:

It brings back the shortened descriptor field to the settings. It was hid because it was not being used on transactions. The usage of the statement descriptor in transactions is being addressed in this PR as well.

For non-UPE transactions, the regular statement descriptor is being added to the payment intents already, here the shortened descriptor is being included to take place on card payments, if enabled. For UPE transactions none of the descriptors were being added, this is also being addressed in this PR.

Another minor change included in this PR is the statement descriptor placeholder. It will be shown when the regular statement descriptor field is cleared. The purpose is to turn what value is going to be set in the payment intent more explicit for the merchant.
<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for the PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

## Testing instructions

![image](https://user-images.githubusercontent.com/10233985/148128538-5f32ed77-b564-44c7-a34d-6a29e8d154c3.png)

### Statement descriptor placeholder

1. Go to the [Stripe settings page](http://localhost:8082/wp-admin/admin.php?page=wc-settings&tab=checkout&section=stripe&panel=settings).
2. Empty the `Full bank statement` field.
3. Ensure the input's placeholder matches what's set in `Statement descriptor` in your [Stripe's account settings](https://dashboard.stripe.com/settings/account).
4. In the preview, the rightmost box should also display the same value of the placeholder.

### Descriptors set are sent to the payment intent

There's a vast amount of combinations for this test, please try to test as much as you can:
- UPE enabled and disabled.
- Classic checkout and Blocks checkout.
- Card, non-card payment method and express checkout.
- `Add customer order number to the bank statement` enabled and disabled.

Also consider the following:
1. When no `Full bank statement` and `Shortened customer bank statement` are provided, the payment intent should either contain no statement descriptor or have the Stripe account's statement descriptor set.
2. When `Full bank statement` is provided and `Add customer order number to the bank statement` is disabled, the payment intent should contain the `Full bank statement`'s value in the statement descriptor regardless the payment method chosen.
3. In card transactions (express checkout included), when `Add customer order number to the bank statement` is enabled and `Shortened customer bank statement` is provided, the payment intent should contain the `Shortened customer bank statement`'s value concatenated with the order number in the statement descriptor, as shown in the preview.
- There's a edge case for (2): If the payment method is not card or SEPA, and the transaction is completed through the classic checkout, the payment intent won't have the statement descriptor set (Stripe request errors if we try to add it for other payment methods).

**The test:**
1. Go to the [shop page](http://localhost:8082/shop/) and add a product to the cart.
2. Go to checkout and complete the order using the scenarios described above.
3. Go to [Payments](https://dashboard.stripe.com/test/payments) in the Stripe dashboard, find the payment you've just made and click on it.
4. Check whether the statement descriptor value (under Payment details section) match the expected value according to the scenario being tested.

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Please follow the following guidelines when writing testing instructions:

- Include screenshots if there is no similar flow in the critical flows: https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Critical-flows
- Assume instructions will be copied over to the Release Testing Instructions wiki page: https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions
- Assume instructions will be followed by external testers.
- Assume tester does not have intimate knowledge of Stripe.
-->

---

-   [x] Covered with tests (or have a good reason not to test in description ☝️)
-   [ ] Added changelog entry **in both** `changelog.txt` and `readme.txt` (or does not apply)
-   [ ] Tested on mobile (or does not apply)

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)

cesarcosta99 committed:
Revert hide shortened statement descriptor

cesarcosta99 committed:
Add account's own statement descriptor placeholder to settings fields

Adding the placeholder in the statement descriptor fields we have in the settings (full and shortened).

cesarcosta99 committed:
Tests: add statement descriptor placeholders test

cesarcosta99 committed:
Add shortened statement descriptor helper

cesarcosta99 committed:
Add shortened descriptor logic to non-UPE payment requests

cesarcosta99 committed:
Add descriptor logic to UPE payment requests

cesarcosta99 committed:
Remove shortened descriptor placeholder

It cannot contain a placeholder because we can't replace the prefix. We're using the shortened descriptor and concatenating with the order ID, generating a full descriptor without having to set up a suffix for it.

cesarcosta99 committed:
Tests: Remove shortened descriptor placeholder assertion

cesarcosta99 committed:
Update descriptor preview to include its placeholder in case it's blank

cesarcosta99 committed:
Tests: Add statement descriptor to the request on UPE gateway tests

cesarcosta99 committed:
Update shortened statement descriptor helper

Make the order parameter optional to avoid compatibility errors.

dechov commented on includes/class-wc-stripe-helper.php:
```diff
@@ -443,6 +443,25 @@ public static function get_order_by_setup_intent_id( $intent_id ) {
+	public static function get_shortened_statement_descriptor( $statement_descriptor = '', $order = null ) {
+		$statement_descriptor = self::clean_statement_descriptor( $statement_descriptor );
+
+		if ( method_exists( $order, 'get_id' ) && ! empty( $order->get_id() ) ) {
+			$statement_descriptor = $statement_descriptor . '* #' . $order->get_id();
```
I think we'll want `$order->get_order_number()` – the ID is the post ID corresponding to the stored order but order number _can_ be different (being filterable) and is what's presented to the customer (e.g. see [template](https://github.com/woocommerce/woocommerce/blob/747cb6b7184ba9fdc875ab104da5839cfda8b4be/plugins/woocommerce/templates/myaccount/orders.php#L49)).

cesarcosta99 replied:
Good catch, fixed in eaa6b2b.

dechov commented on client/settings/payments-and-transactions-section/__tests__/index.test.js:
```diff
@@ -10,6 +11,10 @@ import {
 
+jest.mock( '../../../data/account', () => ( {
+	useAccount: jest.fn(),
+} ) );
```
These paths (and in the component module) can be made independent of nesting level, see for instance:
https://github.com/woocommerce/woocommerce-gateway-stripe/blob/13b254d7e16f5123c3e1110f48c78ddda586adab/client/components/payment-method-capability-status-pill/__tests__/payment-method-capability-status-pill.test.js#L4-L9

cesarcosta99 replied:
Done in b106b0f.

dechov replied:
Thanks – not a big deal, but it's a similar case in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2263/commits/289285d5cdded32704b0999c7dac8db7e5211e24#diff-b6636d9b7c977e87e0fa162a64a0adb5a0cff40145488db12d4234f0e920ba03R5

cesarcosta99 replied:
Done in 8408fcd.

dechov commented on includes/class-wc-stripe-helper.php:
```diff
@@ -443,6 +443,25 @@ public static function get_order_by_setup_intent_id( $intent_id ) {
+		$statement_descriptor = self::clean_statement_descriptor( $statement_descriptor );
+
+		if ( method_exists( $order, 'get_id' ) && ! empty( $order->get_id() ) ) {
+			$statement_descriptor = $statement_descriptor . '* #' . $order->get_id();
```
Do I understand correctly that we're doing this concatenation (joined by `* `) ourselves instead of passing the suffix (as described in [this doc section](https://stripe.com/docs/statement-descriptors#dynamic)) because that is the only way for us to manage the shortened descriptor (prefix) outside of the Stripe Dashboard?

cesarcosta99 replied:
Yes, that's correct. If we pass the suffix, it will be concatenated with the shortened descriptor from the Stripe account and not from the value set in the settings. If we try to pass both `statement_descriptor` and the `statement_descriptor_suffix` the `statement_descriptor` value is completely ignored, besides, the "prefix" is not customizable at all.

dechov commented on tests/phpunit/test-class-wc-stripe-upe-payment-gateway.php:
```diff
@@ -190,6 +190,7 @@ public function test_process_payment_returns_valid_response() {
 			'description' => $description,
 			'customer'    => $customer_id,
 			'metadata'    => $metadata,
+			'statement_descriptor' => null,
```
No big deal but I believe by convention the other values should be indented further here (and below).

cesarcosta99 replied:
Fixed in 2702632.

dechov commented on includes/abstracts/abstract-wc-stripe-payment-gateway.php:
```diff
@@ -1233,6 +1237,10 @@ public function generate_create_intent_request( $order, $prepared_source ) {
 		}
 
+		if ( isset( $full_request['statement_descriptor_suffix'] ) ) {
+			$request['statement_descriptor_suffix'] = $full_request['statement_descriptor_suffix'];
```
Is this here so that the param can be injected via a filter? Do we know how this would interact with the (concatenated) full descriptor we are passing?

cesarcosta99 replied:
Thanks for raising this. This is actually from the very beginning when I was testing out the prefix + suffix case. As explained above, it won't work so I removed it in a8780a4.

dechov commented on includes/class-wc-stripe-helper.php:
```diff
@@ -443,6 +443,25 @@ public static function get_order_by_setup_intent_id( $intent_id ) {
+		$statement_descriptor = self::clean_statement_descriptor( $statement_descriptor );
+
+		if ( method_exists( $order, 'get_id' ) && ! empty( $order->get_id() ) ) {
+			$statement_descriptor = $statement_descriptor . '* #' . $order->get_id();
```
Are we concerned at all about the concatenated statement descriptor being too long? If the static part is up to 13 characters, that leaves 9 for the order number (if that's what we use instead of ID), which seems like plenty for sequential IDs but might not be enough for all cases. Should we truncate it to avoid an error?

cesarcosta99 replied:
Good point, done in 9446e3d.

dechov reviewed (commented):
I have yet to test it but left a few initial comments – the rest of the code all LGTM so far.

cesarcosta99 committed:
Update shortened statement descriptor helper

Replace  with  because they can be different in case they're filtered.

cesarcosta99 committed:
Fix import path

cesarcosta99 committed:
Fix linting

cesarcosta99 committed:
Remove statement descriptor suffix logic

cesarcosta99 committed:
Update shortened statement descriptor helper

Truncate the descriptor to 22 characters to avoid edge cases and possible errors.

cesarcosta99 committed:
Merge branch 'develop' into fix/2205-statement-descriptor-fields

cesarcosta99 committed:
Fix import path

cesarcosta99 committed:
Merge branch 'develop' into fix/2205-statement-descriptor-fields

dechov commented on includes/class-wc-stripe-helper.php:
```diff
@@ -443,6 +443,28 @@ public static function get_order_by_setup_intent_id( $intent_id ) {
 	}
 
+	/**
+	 * Sanizite and retrieve a shortened statement descriptor.
```
Typo, should be "sanitize"

cesarcosta99 replied:
Fixed in 18cc7bb.

dechov commented on includes/class-wc-stripe-helper.php:
```diff
@@ -443,6 +443,28 @@ public static function get_order_by_setup_intent_id( $intent_id ) {
+	 * @param WC_Order $order
+	 * @return string $statement_descriptor Final shortened statement descriptor
+	 */
+	public static function get_shortened_statement_descriptor( $statement_descriptor = '', $order = null ) {
```
The name of this function would make me think it retrieves the shortened static prefix – given that this includes the dynamic suffix, maybe we could call this the "dynamic statement descriptor"?

cesarcosta99 replied:
Good call, renamed it in 18cc7bb.

dechov commented on includes/payment-methods/class-wc-stripe-upe-payment-gateway.php:
```diff
@@ -495,17 +495,29 @@ public function process_payment( $order_id, $retry = true, $force_save_source =
+		$short_statement_descriptor            = ! empty( $this->get_option( 'short_statement_descriptor' ) ) ? str_replace( "'", '', $this->get_option( 'short_statement_descriptor' ) ) : '';
+		$is_short_statement_descriptor_enabled = ! empty( $this->get_option( 'is_short_statement_descriptor_enabled' ) ) && 'yes' === $this->get_option( 'is_short_statement_descriptor_enabled' );
+		$descriptor                            = null;
+		if ( 'card' === $selected_upe_payment_type && $is_short_statement_descriptor_enabled && ! empty( $short_statement_descriptor ) ) {
```
If the setting ("Add customer order number to the bank statement") is enabled, I feel like we should still include the dynamic bit even if the "Shortened customer bank statement" value is empty. We could truncate it to 10 characters, or just truncate it to whatever length makes it fit with the suffix?

(Applies to non-UPE condition as well.)

cesarcosta99 replied:
The validation of the settings is built in a way that it won't let the user save a blank "Shortened customer bank statement" if "Add customer order number to the bank statement" is enabled. I can't see any use case where the proposed behavior would take place. Do you have some specific concern for us to fallback to the regular statement descriptor?

dechov replied:
Ah, I see – I believe it let me save an empty one, but I can try again.

dechov replied:
Looks like an unintended consequence of https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2140 since the modified function is used by the validation of both fields?

cesarcosta99 replied:
True enough. I'll submit that todo in the settings milestone so we can remember to remove it at some point. I've included the fallback in 7d8e94e.

dechov replied:
Hm, I'm not too sure about including the whole descriptor as a fallback, since it's not unlikely to be near the limit already, so the order number is likely to be truncated. (Let's say the order number is 673, then the descriptor could be `FULLBANKSTATEMENT* #67`.) Also, this isn't reflected in the Preview, which just shows an empty prefix.

I'm not too concerned about it if this state will be disallowed soon.

dechov reviewed (commented):
Did a series of tests across a range of combinations as noted in the testing instructions, and all worked well! Noticed one case where we're not including the Order ID even when it's checked.

cesarcosta99 committed:
Rename statement descriptor helper

cesarcosta99 committed:
Add fallback on the dynamic descriptor helper

The fallback is being added because currently the field validation has a breach that needs to be worked out and allows the user to save the descriptors blank.

cesarcosta99 committed:
Merge branch 'develop' into fix/2205-statement-descriptor-fields

dechov reviewed (approved):
Looks good, pre-approving – just need to fix conflict and possibly resolve empty prefix case, which seems minor if the validation failure will be restored in the same milestone.

cesarcosta99 committed:
Tweak the dynamic statement descriptor helper

cesarcosta99 committed:
Merge branch 'develop' into fix/2205-statement-descriptor-fields

cesarcosta99 committed:
Tweak the dynamic statement descriptor helper

dechov reviewed (approved):
LGTM!
